src/Compiler/Utilities/Caches.fs

```diff
@@ -0,0 +1,174 @@
+// which is *not* calling `ConcurrentDictionary.ToArray`, but uses a custom one instead (treating it as `ICollection`)
+// this leads to and exception when trying to evict without locking (The index is equal to or greater than the length of the array,
+// or the number of elements in the dictionary is greater than the available space from index to the end of the destination array.)
+// this is casuedby insertions happened between reading the `Count` and doing the `CopyTo`.
+// This solution introduces a custom `ConcurrentDictionary.sortBy` which will be calling a proper `CopyTo`, the one on the ConcurrentDictionary itself.
+module ConcurrentDictionary =
+
+    open System.Collections
+    open System.Collections.Generic
+
+    let inline mkSeq f =
+        { new IEnumerable<'U> with
+            member _.GetEnumerator() = f()
+
+          interface IEnumerable with
+            member _.GetEnumerator() = (f() :> IEnumerator) }
+
+    let inline mkDelayedSeq (f: unit -> IEnumerable<'T>) =
+        mkSeq (fun () -> f().GetEnumerator())
+
+    let inline sortBy ([<InlineIfLambda>] projection) (source: ConcurrentDictionary<_, _>) =
+        mkDelayedSeq (fun () ->
+            let array = source.ToArray()
+            Array.sortInPlaceBy projection array
+            array :> seq<_>)
+
+[<Struct; RequireQualifiedAccess; NoComparison; NoEquality>]
+type CachingStrategy = LRU | LFU
+
+[<Struct; RequireQualifiedAccess; NoComparison>]
+type EvictionMethod = Blocking | Background
+
+[<Struct; RequireQualifiedAccess; NoComparison; NoEquality>]
+type CacheOptions =
+    { MaximumCapacity: int
+      PercentageToEvict: int
+      Strategy: CachingStrategy
+      EvictionMethod: EvictionMethod
+      LevelOfConcurrency: int } with
+    static member Default = { MaximumCapacity = 100; PercentageToEvict = 5; Strategy = CachingStrategy.LRU; LevelOfConcurrency = Environment.ProcessorCount; EvictionMethod = EvictionMethod.Blocking; }
+
+[<Sealed; NoComparison; NoEquality>]
+type CachedEntity<'Value> =
+    val Value: 'Value
+    val mutable LastAccessed: int64
+    val mutable AccessCount: uint64
+
+    new(value: 'Value) = { Value = value; LastAccessed = DateTimeOffset.Now.Ticks; AccessCount = 0UL }
+
+
+[<Sealed; NoComparison; NoEquality>]
+[<DebuggerDisplay("{GetStats()}")>]
+type Cache<'Key, 'Value> (options: CacheOptions) as this =
+
+    let capacity = options.MaximumCapacity + (options.MaximumCapacity * options.PercentageToEvict / 100)
+    let cts = new CancellationTokenSource()
+
+    do
+        if options.EvictionMethod = EvictionMethod.Background then
+            Task.Run(this.TryEvictTask, cts.Token) |> ignore
+
+    let cacheHit = Event<_ * _>()
+    let cacheMiss = Event<_>()
+    let eviction = Event<_>()
+
+    [<CLIEvent>] member val CacheHit = cacheHit.Publish
+    [<CLIEvent>] member val CacheMiss = cacheMiss.Publish
+    [<CLIEvent>] member val Eviction = eviction.Publish
```

Comment:
So, one thing about these is that since so many things are let bindings on type, this will generate init code, and many, if not all members will include a check for init, like `if init@52 then ...`, which is not ideal. 

I might rethink how initialisation is working here.

Reply:
I have fixed it by changing structure a bit, instead of using constructors and type initialiser to use a factory method. Now initialisation is much more straightforward.